tsdate/variational.py

```diff
@@ -477,77 +560,127 @@ def propagate_mutations(
+                    child_cavity,
+                    edge_likelihood,
                 )
             elif fixed[c] and not fixed[p]:
                 parent_message = factors[i, ROOTWARD] * scale[p]
                 parent_delta = 1.0  # hopefully we don't need to damp
                 parent_cavity = posterior[p] - parent_delta * parent_message
                 edge_likelihood = parent_delta * likelihoods[i]
                 child_age = constraints[c, LOWER]
-                mutations_posterior[m] = approx.mutation_rootward_projection(
-                    child_age, parent_cavity, edge_likelihood, min_kl
+                mutations_phase[m], mutations_posterior[m] = rootward_projection(
+                    child_age,
+                    parent_cavity,
+                    edge_likelihood,
                 )
             else:
-                parent_message = factors[i, ROOTWARD] * scale[p]
-                child_message = factors[i, LEAFWARD] * scale[c]
-                parent_delta = 1.0  # hopefully we don't need to damp
-                child_delta = 1.0  # hopefully we don't need to damp
-                delta = min(parent_delta, child_delta)
-                parent_cavity = posterior[p] - delta * parent_message
-                child_cavity = posterior[c] - delta * child_message
-                edge_likelihood = delta * likelihoods[i]
-                mutations_posterior[m] = approx.mutation_gamma_projection(
-                    parent_cavity, child_cavity, edge_likelihood, min_kl
-                )
+                if p == c:  # singleton block with single parent
+                    parent_message = factors[i, ROOTWARD] * scale[p]
+                    parent_delta = 1.0  # hopefully we don't need to damp
+                    parent_cavity = posterior[p] - parent_delta * parent_message
+                    edge_likelihood = parent_delta * likelihoods[i]
+                    child_age = constraints[c, LOWER]
+                    mutations_phase[m], mutations_posterior[m] = \
+                        twin_projection(parent_cavity, edge_likelihood)  # fmt: skip
+                else:
+                    parent_message = factors[i, ROOTWARD] * scale[p]
+                    child_message = factors[i, LEAFWARD] * scale[c]
+                    parent_delta = 1.0  # hopefully we don't need to damp
+                    child_delta = 1.0  # hopefully we don't need to damp
+                    delta = min(parent_delta, child_delta)
+                    parent_cavity = posterior[p] - delta * parent_message
+                    child_cavity = posterior[c] - delta * child_message
+                    edge_likelihood = delta * likelihoods[i]
+                    mutations_phase[m], mutations_posterior[m] = gamma_projection(
+                        parent_cavity,
+                        child_cavity,
+                        edge_likelihood,
+                    )
 
-        return mutations_posterior
+        return np.nan
```

Comment:
I'm not understanding why we return anything here (should it be documented). What was the returned posterior being used for before?